protocols/relay/src/priv_client/handler.rs

```diff
@@ -197,82 +188,64 @@ impl Handler {
             <Self as ConnectionHandler>::OutboundProtocol,
         >,
     ) {
-        let outbound_info = self.wait_for_outbound_stream.pop_front().expect(
-            "got a stream error without a pending connection command or a reserve listener",
-        );
-        match outbound_info {
-            outbound_hop::OutboundStreamInfo::Reserve(mut to_listener) => {
-                let non_fatal_error = match error {
-                    StreamUpgradeError::Timeout => StreamUpgradeError::Timeout,
-                    StreamUpgradeError::NegotiationFailed => StreamUpgradeError::NegotiationFailed,
-                    StreamUpgradeError::Io(e) => {
-                        self.pending_error = Some(StreamUpgradeError::Io(e));
-                        return;
+        let pending_request = self
+            .pending_requests
+            .pop_front()
+            .expect("got a stream error without a pending request");
+
+        match pending_request {
+            PendingRequest::Reserve { mut to_listener } => {
+                let error = match error {
+                    StreamUpgradeError::Timeout => {
+                        outbound_hop::ReserveError::Io(io::ErrorKind::TimedOut.into())
                     }
-                    StreamUpgradeError::Apply(v) => void::unreachable(v),
+                    StreamUpgradeError::Apply(never) => void::unreachable(never),
+                    StreamUpgradeError::NegotiationFailed => {
+                        outbound_hop::ReserveError::Unsupported
+                    }
+                    StreamUpgradeError::Io(e) => outbound_hop::ReserveError::Io(e),
                 };
 
-                if self.pending_error.is_none() {
-                    self.send_error_futs.push(
-                        async move {
-                            let _ = to_listener
-                                .send(transport::ToListenerMsg::Reservation(Err(())))
-                                .await;
-                        }
-                        .boxed(),
-                    );
-                } else {
-                    // Fatal error occurred, thus handler is closing as quickly as possible.
-                    // Transport is notified through dropping `to_listener`.
-                }
-
-                let renewal = self.reservation.failed();
-
-                self.queued_events
-                    .push_back(ConnectionHandlerEvent::NotifyBehaviour(
-                        Event::ReservationReqFailed {
-                            renewal,
-                            error: non_fatal_error,
-                        },
-                    ));
+                self.send_channel_tasks.push(
+                    async move {
+                        let _ = to_listener
+                            .send(transport::ToListenerMsg::Reservation(Err(error)))
+                            .await;
+                    }
+                    .boxed(),
+                );
```

Comment:
Adding a new `Future` to a `FutureUnordered`, which then sends an event into a channel, is effectively the same as an unbounded channel. I would consider an unbounded channel simpler.

Instead of switching to an unbounded channel, how about using a channel with a large capacity (e.g. `1_000`), sending the error into the channel on a best effort way, logging a `warn!` when the channel is full.

Reply:
I can just use `try_send` then, right? The only thing that the async bit does is enforcing backpressure.